This is a WAY better attempt at X-ray mode.

henryxparker opened the pull request.

henryxparker commented on mtags-shared/src/main/scala/scala/meta/internal/pc/InlayHints.scala:
```diff
@@ -123,6 +181,36 @@ object InlayHints {
+    hints: List[BlockInlayHint]
+) {
+  def build: List[BlockInlayHint] = {
+    if (hints.length == 1) {
```
I tried to do this filtering in the `unapply` but I would need the LSP positions and when I tried to access them at that point the tests all failed with a scary "compiler broken!! :c" message so I decided to just do it here

henryxparker commented on tests/cross/src/test/scala/tests/pc/InlayHintsSuite.scala:
```diff
@@ -16,6 +16,302 @@ class InlayHintsSuite extends BaseInlayHintsSuite {
+      |val foo: Foo = ???
+      |
+      |val thing1: Bar = bar
+      |  .bar/*: Bar<<(2:8)>>*/
```
I figured it makes sense to show the types on one-liners if you indent the call. This can easily be removed if people don't like it

henryxparker replied:
I decided to not do this anymore. Theres just too many one liners that look bad with it

henryxparker commented on tests/cross/src/test/scala/tests/pc/InlayHintsSuite.scala:
```diff
@@ -16,6 +16,302 @@ class InlayHintsSuite extends BaseInlayHintsSuite {
+      |val thingy: Bar = foo
+      |  .bar/*  : Bar<<(6:8)>>*/
+      |  .foo()/*: Foo<<(2:8)>>*/
+      |  .bar/*  : Bar<<(6:8)>>*/
```
getting all of the inlay hints for a block justified alongside each other feels so good. Please don't make me get rid of it orz

tgodzik commented on tests/unit/src/test/resources/inlayHints/example/ImplicitConversions.scala:
```diff
@@ -11,7 +11,7 @@ class ImplicitConversions {
 
   // extension methods
   /*augmentString<<scala/Predef.augmentString().>>(*/message/*)*/
-    .stripSuffix(/*suffix = */"h")
```
We can leave it be for now, it's anyway under a flag. This shouldn't be to hard to change, basically we would need to check if we are at the last select.

henryxparker commented on mtags/src/main/scala-2/scala/meta/internal/pc/PcInlayHintsProvider.scala:
```diff
@@ -102,16 +103,23 @@ final class PcInlayHintsProvider(
   ): InlayHints = {
-    val inlayHints = collectDecorations(tree, acc)
-    tree.children.foldLeft(inlayHints)(traverse(_, _))
+    val inlayHints = collectDecorations(tree, parent, acc)
```
Yes, if we want access to the parent tree, which we need for a couple checks in the `unapply`

zielinsky commented on mtags-shared/src/main/scala/scala/meta/internal/pc/InlayHints.scala:
```diff
@@ -32,6 +33,59 @@ case class InlayHints(
+      ) { (ihb: InlayHintBlock) =>
+        val newLevel = pos.getEnd.getCharacter
+
+        val indentLevel = if (newLevel > ihb.indentLevel) {
```
is just math.max

zielinsky commented on mtags-shared/src/main/scala/scala/meta/internal/pc/InlayHints.scala:
```diff
@@ -123,6 +186,36 @@ object InlayHints {
+        val labels =
+          if (naiveIndent <= 0) hint.labels
+          else LabelPart(" ".repeat(naiveIndent)) :: hint.labels
+        BlockInlayHint(
```
We could use hint.copy(labels = labels) here - for example if someone adds new fields to BlockInlayHint, copy() will handle them automatically

tgodzik reviewed (commented):
Also, maybe we would want a similar name as Intellij so that it's more clear to people?

henryxparker commented:
> @henryxparker could you take a look at the failing tests?

Hey, so there MUST have been a bug or some kind of change introduced since I started working on this, specifically in regards to how `Apply` works. I might need help locating it.
The `a @ Apply(inner, _)` unapply in my code is never matching, so I did what I thought would help debug and removed some of the boolean filters to see what is unexpectedly false. The problem is, I ended up taking out ALL the booleans and it STILL never matched, or so I thought.  
I was using a basic `println` in the apply to verify it's getting called, cuz they appear in the tests. It was not printing that line ever even with no boolean filters, so I checked adding a `println` to the select and that one appeared just fine. I tried adding a `println` to the apply of a different hint type and That one also appeared just fine. So in frustration, I added this:
```scala
tree match {
          case Apply(_, _) => println("got to an apply")
}
```
I added that right above the other pattern match hoping it would print something. BUT this just causes all the other hint types to silently fail! WTF?! So I think the `Apply.unapply` was broken in some way after I started working on it, and rebasing my code introduced that bug and caused them to fail.

tgodzik commented:
> I assume 2.11 does something different with for comprehensions? I'm not sure how to resolve that failing cross-test. If I were in charge I would just remove that test because I'm lazy (also who uses 2.11 anymore?), but I would understand the argument that we shouldn't have different behavior between versions. If anyone has an idea as to why it might be different I would love to hear it

It's fine if things do not work perfectly for 2.11, if there is a big issue we can even disable the feature for that version. Though I see there is just one more test failing and it' the same for 2.11 and later. Otherwise the PR looks really good, will you have time to look at that last test?

tgodzik commented:
The remaining error is not related, I will merge this before the next release

tgodzik reviewed (approved):
LGTM! Great work!
